### PR TITLE
Refactor tests that cause warnings in CI

### LINKED ISSUE #50337: roslyn-ci has test publishing warnings due to too many subresults in test case
I wanted to make note of this warning in the [release branch run](https://dev.azure.com/dnceng/public/_build/results?buildId=938999&view=results) after this PR was merged:

```log
##[warning]No of subresults in test case: Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableContextTests.AnalyzeMethodsInEnabledContextOnly_SpeculativeSemanticModel are more than supported limit 1000, truncating remaning ones.
##[warning]No of subresults in test case: Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableContextTests.AnalyzeMethodsInEnabledContextOnly_01 are more than supported limit 1000, truncating remaning ones.
```

I'm not certain but this indicates a possibility that if these tests start failing in the future that not all the results will be shown in the AzDo UI. Ideally we could make some modification that prevents the warnings from showing in the AzDo run page, either by refactoring the tests or suppressing the warning.

_Originally posted by @RikkiGibson in https://github.com/dotnet/roslyn/issues/49953#issuecomment-754923927_

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #50337

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs [hunk 1]
```diff
@@ -454,40 +474,41 @@ public override string ToString()
             }
         }
 
-        public static IEnumerable<object[]> AnalyzeMethodsInEnabledContextOnly_01_Data()
+        // AnalyzeMethodsInEnabledContextOnly_01_Data is splitted due to https://github.com/dotnet/roslyn/issues/50337
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:477]: "split"?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs [hunk 2]
```diff
@@ -21,6 +21,26 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
 {
     public class NullableContextTests : CSharpTestBase
     {
+        private static readonly NullableDirectives[] s_nullableDirectives = new[]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:24]: >s_nullableDirectives [](start = 53, length = 20)

Consider moving this immediately before `AnalyzeMethodsInEnabledContextOnly_01_Data1()`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs [hunk 3]
```diff
@@ -496,7 +517,8 @@ public static IEnumerable<object[]> AnalyzeMethodsInEnabledContextOnly_01_Data()
         }
 
         [Theory]
-        [MemberData(nameof(AnalyzeMethodsInEnabledContextOnly_01_Data))]
+        [MemberData(nameof(AnalyzeMethodsInEnabledContextOnly_01_Data1))]
+        [MemberData(nameof(AnalyzeMethodsInEnabledContextOnly_01_Data2))]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:521]: Does this address the issue? Don't we still have the same number of test results per test?

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:521]: @cston I'm waiting for CI to check. But I think while it's the same number per *test*, it's not the same number per *test case*. The warning is about "subresults in *test case*". But CI will say its decision :smile:

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:521]: If the same issue exists, consider having two tests that share a common helper in each case.

```csharp
[MemberData(nameof(AnalyzeMethodsInEnabledContextOnly_01_Data1)]
public void AnalyzeMethodsInEnabledContextOnly_01A(NullableContextOptions? projectContext, NullableDirectives classDirectives, NullableDirectives methodDirectives)
{
    AnalyzeMethodsInEnabledContextOnly_01_Execute(projectContext, classDirectives, methodDirectives);
}

[MemberData(nameof(AnalyzeMethodsInEnabledContextOnly_01_Data2)]
public void AnalyzeMethodsInEnabledContextOnly_01B(NullableContextOptions? projectContext, NullableDirectives classDirectives, NullableDirectives methodDirectives)
{
    AnalyzeMethodsInEnabledContextOnly_01_Execute(projectContext, classDirectives, methodDirectives);
}

private void AnalyzeMethodsInEnabledContextOnly_01_Execute(NullableContextOptions? projectContext, NullableDirectives classDirectives, NullableDirectives methodDirectives)
{
    ...
}
```

---
In reply to: [589068480](https://github.com/dotnet/roslyn/pull/51730#discussion_r589068480) [](ancestors = 589068480)

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:521]: @cston You're correct, the warnings are still there 😕 

image

I don't have a good solution in mind. Duplicating the exact same test doesn't seem to be a good option. I'll close and let someone else come with a good solution.

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs:521]: @cston Oh, I had the PR tab open and it didn't show your previous reply when I responded. I'll reopen and apply your change. Thanks.

**Review by cston — APPROVED** [review]

**Comment by Youssef1313** [comment]
Warnings are fixed :tada:

image

Thanks @cston

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Thanks @Youssef1313!

**Pull request merged**